### PR TITLE
Sql editor

### LINKED ISSUE #7759: Add SQL editor for data panel
- [x] Add switcher between data content or sql panel.
- [x] Add SQL editor.
- [x] Add SQL parser
- [x] Don't let edit SQL if an analysis is applied.
- [ ] ~~Add data panel with new stats~~ (cc @fdansv). #7815
- [ ] ~~Retro compatibility between layer sql and query schema model query.~~ #7843

### PR DESCRIPTION
**Pull request by nobuti**
This PR closes #7759.

- The main view is a placeholder, the stats integration will happen in #7815.
- The editor turns read only if an analysis is applied.
- The sql parser shows the query errors, but we are not able to show the line number.

Also:
- [x] we need to update the layer definition after a custom sql is applied, but we need to discuss how we want to perform the action
- [x] we need to add retro compatibility with the current editor.

In addition, the infobox's styles need some fixes when it is shown above a dark background. Its background has an opacity, so the color is not accurate. cc @piensaenpixel.

screen shot 2016-06-14 at 10 06 31

cc @xavijam @javisantana @viddo

### CODE REVIEW

**Comment by xavijam** [comment]
One thing, if SQL editor is disabled, apply and undo-redo buttons should not be visible.

File: lib/assets/javascripts/cartodb3/data/query-schema-model.js [hunk 1]
```diff
@@ -62,8 +62,12 @@ module.exports = Backbone.Model.extend({
     );
 
     opts.method = this._httpMethod();
-    opts.error = function () {
-      this.set('status', 'unavailable');
+    opts.error = function (model, response) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/query-schema-model.js]: I'm not sure if we should "store" the error message in the query-schema-model. But if we do so, we should clean it when the response is 🆗 .

File: lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror-view.js [hunk 2]
```diff
@@ -49,22 +56,34 @@ module.exports = CoreView.extend({
   _bindEvents: function () {
     var self = this;
     this.editor.on('change', function (editor, changed) {
-      self.model.set('content', self.getContent(), { silent: true });
+      self.model.set('content', self.getContent(), {silent: true});
     });
 
-    this.model.bind('change:content', function () {
+    this.model.on('change:content', function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror-view.js]: with not `bind`? do you prefer `on`? mania?

File: lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror.tpl [hunk 3]
```diff
@@ -2,6 +2,6 @@
   <textarea class="js-editor"><%- content %></textarea>
 </div>
 <div class="CodeMirror-console">
-  CMD + S to apply your styles.
+  <%- tip %>
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror.tpl]: 👍 , have we "internationalized" these texts?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/code-mirror-error.tpl [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<ul class="CodeMirror-error">
+  <li class="CodeMirror-errorMessage u-lSpace--xl u-rSpace--xl">
+    Syntax error: <span><%- message %></span>
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/code-mirror-error.tpl]: Let's move strings to `en.json` file :).

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/code-mirror-error.tpl [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+  </li>
+  <li class="CodeMirror-errorDocs">
+    <a href="https://docs.cartodb.com/cartodb-platform/sql-api/" target="_black">DOCS</a>
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/code-mirror-error.tpl]: Same ^^^

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-sql-view.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  },
+
+  render: function () {
+    this.clearSubViews();
+    this.$el.empty();
+    this._initViews();
+    return this;
+  },
+
+  _initBinds: function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-sql-view.js]: not needed for the moment, right?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        message: error
+      };
+    });
+  },
+
+  _afterSaveSql: function () {
+    var content = this._codemirrorModel.get('content');
+    this._querySchemaModel.set('query_errors', []);
+    this._layerDefinitionModel.save({
+      sql_custom: true,
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: I feel sql_custom doesn't exist right now in the layer definition, not sure if we should follow same "technique" we use in the style panel... _thinking_

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+      sql: content
+    });
+  },
+
+  _hasAnalysisApplied: function () {
+    return this._layerDefinitionsCollection.at(1).getAnalysisDefinitionNodeModel().get('type') !== 'source';
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: Could we use this @viddo?

File: lib/assets/javascripts/cartodb3/data/query-schema-model.js (hunk 1 above)
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/data/query-schema-model.js]: We are injecting the error option, so the error is set here, but the success callback is managed by a high order view. There the query errors are reset. Not sure if we should change that.

File: lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror.tpl (hunk 3 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror.tpl]: Auto-answering, yes, 🎉 .

**Comment by xavijam** [comment]
Minor things :), 👍

File: lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror-view.js (hunk 2 above)
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror-view.js]: mania for sure. I'm used to see `bind` as a context related utility, and `on` as event related function. I'll change it.

File: lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror.tpl (hunk 3 above)
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror.tpl]: Yes.

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js [hunk 9]
```diff
@@ -161,12 +161,15 @@ module.exports = CoreView.extend({
         });
       },
       createActionView: function () {
-        return new UndoButtons({
-          trackModel: self._sqlModel,
-          editorModel: self._editorModel,
-          applyButton: true,
-          onApplyClick: self._saveSQL.bind(self)
-        });
+        var is_analysis = self._hasAnalysisApplied();
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: Let's try to create local variables in camelCase buddy :).

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js (hunk 8 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: Why always get the first layer (`this._layerDefinitionsCollection.at(1)`)?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js (hunk 7 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: The query should be updated on the analysis-source model, not the layer-definition. The sql property there is only there due to backward compatibility for a viz that have not yet persisted its analysis (e.g. a new viz created through old editor). We should avoid duplicated state in various locations.

File: lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror-view.js (hunk 2 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/components/code-mirror/code-mirror-view.js]: +1, and its also the [official Backbone events methods](http://backbonejs.org/#Events)

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js (hunk 8 above)
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: @viddo how should be the proper way?

Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/data/data-view.js]: From what I can see/understand, shouldn't it be `this._layerDefinitionModel`?

**Comment by nobuti** [comment]
🚀

**Pull request merged**